### PR TITLE
Use analytic kl divergence in `TraceEnum_ELBO`

### PR DESCRIPTION
**Pull request by ordabayevy**
This PR adds support for use of analytic kl divergence when possible to `TraceEnum_ELBO`.

### CODE REVIEW

File: numpyro/infer/elbo.py [hunk 1]
```diff
@@ -932,10 +932,26 @@ def single_particle_elbo(rng_key):
...
                     assert len(group_names) == 1
                     name = next(iter(group_names))
-                    cost = model_trace[name]["log_prob"]
-                    scale = model_trace[name]["scale"]
-                    deps = model_deps[name]
-                    dice_factors = [guide_trace[key]["log_measure"] for key in deps]
+                    try:
+                        cost = -kl_divergence(
+                            guide_trace[name]["fn"], model_trace[name]["fn"]
+                        )
```
Comment (ordabayevy) [inline numpyro/infer/elbo.py:938]: Maybe move `try/except` logic to `get_importance_trace_enum` to avoid wasteful computation of `log_prob`s?

**Review by fehiepsi — COMMENTED** [review]
So this one computes KL before marginalizing out discrete variables? I guess the order does not matter.

File: numpyro/infer/elbo.py (hunk 1 above)
Comment (martinjankowiak) [reply numpyro/infer/elbo.py]: are you sure this is correct? what are the conditions under which you can compute the KL? note the conditions i wrote down for the mean field elbo:

```
The precise condition is that for every
    latent variable `z` in the guide, its parents in the model must not include
    any latent variables that are descendants of `z` in the guide. Here
    'parents in the model' and 'descendants in the guide' is with respect
    to the corresponding (statistical) dependency structure. For example, this
    condition is always satisfied if the model and guide have identical
    dependency structures.
```

File: numpyro/infer/elbo.py [hunk 2]
```diff
@@ -728,43 +730,72 @@ def get_importance_trace_enum(model, guide, args, kwargs, params, max_plate_nest
...
-                    intermediates = site["intermediates"]
+                value = site["value"]
+                intermediates = site["intermediates"]
+                dim_to_name = site["infer"]["dim_to_name"]
+
+                # compute log factor
+                if (
+                    is_model
+                    and (name in guide_trace)
+                    and model_deps[name].isdisjoint(sum_vars | guide_desc[name])
```
Comment (ordabayevy) [inline numpyro/infer/elbo.py:746]: @martinjankowiak can you check this? I think `model_deps[name].isdisjoint(guide_desc[name])` should take care of the condition for analytic kl now.

File: numpyro/infer/elbo.py [hunk 3]
```diff
@@ -728,43 +730,72 @@ def get_importance_trace_enum(model, guide, args, kwargs, params, max_plate_nest
...
+                        if intermediates:
+                            log_prob = site["fn"].log_prob(value, intermediates)
+                        else:
+                            log_prob = site["fn"].log_prob(value)
+                        site["log_prob"] = to_funsor(
+                            log_prob, output=funsor.Real, dim_to_name=dim_to_name
+                        )
+                elif not is_model and (model_trace[name].get("kl", None) is not None):
+                    # skip logq computation if analytic kl was computed
+                    pass
```
[the remark this answers is not in the document]
Comment (ordabayevy) [reply numpyro/infer/elbo.py:765]: @fehiepsi IIUC your question it doesn't compute KL for sites that are being contracted in the model.

File: numpyro/infer/elbo.py (hunk 2 above)
Comment (martinjankowiak) [reply numpyro/infer/elbo.py:746]: can you remind me of some of the context? does this PR support both model-side and guide-side enumeration? what do i need to know about what's happening on the discrete side?

Comment (ordabayevy) [reply numpyro/infer/elbo.py:746]: Yes, it supports both model-side and guide-side enumeration. It also uses provenance tracking to get the dependency structure of terms in the model on model-enumerated sites. Then this info is used to partition model terms into uncontracted terms and terms that are being contracted together. Then each cost term from model/guide is integrated wrt guide-enumerated sites.

Comment (martinjankowiak) [reply numpyro/infer/elbo.py:746]: do we have any tests that cover the KL branch specifically? more generally what tests do we have to cover the case with continuous latents + TraceEnum?

suppose the model is a K-component gaussian mixture. for model side-enumeration we can't compute the KL because we don't have the formula. for guide-side enumeration with a K-component gaussian mixture guide, we can compute both (i.e. one continuous and one discrete) KL divergences, right?

can we check these two simple cases somehow?

is it the case that model-side enumerations makes most KL divergence computations intractable?

Comment (ordabayevy) [reply numpyro/infer/elbo.py:746]: I will try to write some tests.

File: numpyro/infer/elbo.py [hunk 4]
```diff
@@ -892,32 +923,26 @@ def single_particle_elbo(rng_key):
 
             seeded_model = seed(model, model_seed)
             seeded_guide = seed(guide, guide_seed)
+            model_deps, guide_deps = get_nonreparam_deps(
+                seeded_model, seeded_guide, args, kwargs, param_map
+            )
 
-            model_trace, guide_trace = get_importance_trace_enum(
+            seeded_model = seed(model, model_seed)
+            seeded_guide = seed(guide, guide_seed)
```
[the remark this answers is not in the document]
Comment (fehiepsi) [reply numpyro/infer/elbo.py:932]: yeah, it's hard to reason about this :(

File: test/infer/test_gradient.py [hunk 5]
```diff
@@ -133,3 +133,216 @@ def actual_loss_fn(params_raw):
...
+    probs_z2 = jnp.array([0.5, 0.5])
+    probs_z1 = jnp.array([[0.4, 0.6], [0.6, 0.4]])
+    probs_z3 = jnp.array([[[0.4, 0.6], [0.1, 0.9]], [[0.7, 0.3], [0.2, 0.8]]])
+    z2 = pyro.sample("z2", dist.Categorical(probs_z2))
+    z1 = pyro.sample("z1", dist.Categorical(probs_z1[z2]))
+    pyro.sample("z3", dist.Categorical(probs_z3[z1, z2]))
+
+
+@config_enumerate
+def kl_model_3_z2z3(params):
```
Comment (ordabayevy) [inline test/infer/test_gradient.py:188]: For ease of reading the code the suffix of model names has latent sites to which analytic KL can be applied.

**Review by ordabayevy — COMMENTED** [review]
@martinjankowiak @fehiepsi
- Added tests that compare analytic kl to enumeration. In both cases exact solution is obtained so it is easy to compare. Feel free to add more tests.
- Added `config_kl` function that takes as an argument `sites` set to which analytic kl should be applied
- If `infer={"kl": "analytic"} but analytic kl cannot be applied then an error is raised. So it is strict in a sense that analytic kl is always applied when configured as such.

**Review by fehiepsi — APPROVED** [review]
Thanks, Yerdos. LGTM pending comments from @martinjankowiak.

**Review by martinjankowiak — COMMENTED** [review]
@ordabayevy looks great left some comments!

can we add some tests which also contain continuous latent variables?
for example some of the GMMs i mentioned.

the simplest case is just to check that
```
def model():
    pyro.sample("z0", dist.Normal(0.0, 1.0))
    probs_z1 = jnp.array([0.5, 0.5])
    pyro.sample("z1", dist.Categorical(probs_z1))
```
gives the correct ELBO when computed using a similarly structure guide.

File: test/infer/test_gradient.py [hunk 6]
```diff
@@ -133,3 +133,236 @@ def actual_loss_fn(params_raw):
...
+        (kl_model_3_z2z3, set(["z2", "z3"]), True),
+        (kl_model_4_z3, set(["z1", "z2", "z3"]), False),
+        (kl_model_4_z3, set(["z3"]), True),
+        (kl_model_5_z2, set(["z1", "z2", "z3"]), False),
+        (kl_model_5_z2, set(["z2"]), True),
+        (kl_model_6_, set(["z1", "z2", "z3"]), False),
+        (kl_model_6_, set(), True),
+        (kl_model_7_z3, set(["z1", "z2", "z3"]), False),
+        (kl_model_7_z3, set(["z3"]), True),
+    ][-2:],
```
[the remark this answers is not in the document]
Comment (ordabayevy) [reply test/infer/test_gradient.py]: Done

File: test/infer/test_gradient.py [hunk 7]
```diff
@@ -133,3 +133,236 @@ def actual_loss_fn(params_raw):
...
+    elbo = infer.TraceEnum_ELBO()
+
+    # Expected grads averaged over num_particles
+    def expected_loss_fn(params_raw):
+        params = jax.tree_util.tree_map(transform, params_raw)
+        return elbo.loss(random.PRNGKey(0), {}, model, guide, params)
+
+    expected_loss, expected_grads = jax.value_and_grad(expected_loss_fn)(params_raw)
+
+    # Grads based on exact integration
```
[the remark this answers is not in the document]
Comment (ordabayevy) [reply test/infer/test_gradient.py]: In both cases integration is exact. Expected values are based on enumeration. Actual value are based on the mix of analytic kl and enumeration. Fixed the comments accordingly.

File: numpyro/handlers.py [hunk 8]
```diff
@@ -221,7 +221,7 @@ def process_message(self, msg):
             if guide_msg["type"] != "sample" or guide_msg["is_observed"]:
                 raise RuntimeError(f"Site {name} must be sampled in trace.")
             msg["value"] = guide_msg["value"]
-            msg["infer"] = guide_msg["infer"]
+            msg["infer"] = guide_msg["infer"].copy()
```
Comment (ordabayevy) [inline numpyro/handlers.py:224]: @martinjankowiak @fehiepsi  Is it okay to copy the `infer` dictionary? Can that possibly mess up other inference algorithms?

I had to create a copy here because computing `dim_to_name` for the site in the model overwrites it in the guide as well. But there can be cases like in new tests in this PR when the guide's `dim_to_name` is different than the one in the model.

File: numpyro/contrib/funsor/infer_util.py [hunk 9]
```diff
@@ -81,6 +81,39 @@ def config_fn(site):
     return infer_config(fn, config_fn)
 
 
+def config_kl(fn=None, sites=None):
+    """
+    Configures analytic kl for all sites in ``sites`` in a NumPyro model.
```
[the remark this answers is not in the document]
Comment (ordabayevy) [reply numpyro/contrib/funsor/infer_util.py]: Done

**Comment by ordabayevy** [comment]
@martinjankowiak I'll work on adding more tests that have continuous latent variables.

**Comment by martinjankowiak** [comment]
> "How would we know if it can be computed in principle?" 
i just mean that the graph dependency structure doesn't disallow integration. so maybe a particular site can be integrated out (in principle) if both sites are normal or if both are studentt but we only have the `kl(normal, normal)` pattern in our list of analytic expressions

File: numpyro/handlers.py (hunk 8 above)
Comment (fehiepsi) [reply numpyro/handlers.py:224]: I think it is fine.

**Comment by ordabayevy** [comment]
@martinjankowiak added a test for a model with continuous latent variable. Does it look alright to you?

> i just mean that the graph dependency structure doesn't disallow integration. so maybe a particular site can be integrated out (in principle) if both sites are normal or if both are studentt but we only have the kl(normal, normal) pattern in our list of analytic expressions

Hmm.. not sure how that would be implemented.

File: test/infer/test_gradient.py [hunk 10]
```diff
@@ -366,3 +366,90 @@ def actual_loss_fn(params_raw):
 
             assert_equal(actual_loss, expected_loss, prec=1e-5)
             assert_equal(actual_grads, expected_grads, prec=1e-5)
+
+
+def kl_model_8_z1z2(params):
```
Comment (martinjankowiak) [inline test/infer/test_gradient.py:371]: can we explicitly compare the corresponding ELBO with mean field guide directly to something like `kl_div(Cat, Cat) + kl_div(Norm, Norm)` in the test?

File: test/infer/test_gradient.py [hunk 11]
```diff
@@ -366,3 +366,90 @@ def actual_loss_fn(params_raw):
...
+    probs_z2 = jnp.array([0.0, 0.0])
+    z1 = pyro.sample("z1", dist.Categorical(probs_z1))
+    pyro.sample("z2", dist.Normal(probs_z2[z1], 1.0))
+
+
+@pytest.mark.parametrize(
+    "model",
+    [kl_model_8_z1z2, kl_model_9_z1z2],
+)
+def test_analytic_kl_continuous(model):
```
Comment (martinjankowiak) [inline test/infer/test_gradient.py:398]: i'm not entirely sure i understand what this testing. can we include a branch with score function gradients for the discrete latient variable, i.e. no enumeration?

**Comment by ordabayevy** [comment]
@martinjankowiak added new tests:

- models in `test_analytic_kl_2` and `test_analytic_kl_3` have a mixture of enumerated, non-reparam, and reparam (kl) sites but with different dependency structures
- `test_analytic_kl_4` tests `handlers.scale` and `plate`.

File: test/infer/test_gradient.py [hunk 12]
```diff
@@ -368,88 +369,255 @@ def actual_loss_fn(params_raw):
...
         }
-        return elbo.loss(random.PRNGKey(0), {}, model, guide, params)
+        kl_z1 = kl_divergence(
+            dist.Categorical(params["probs_z1"]),
+            dist.Categorical(jnp.array([0.5, 0.5])),
+        )
+        kl_z2 = kl_divergence(
+            dist.Normal(params["probs_z2"]), dist.Normal(jnp.array(0.0))
+        )
+        return scale * (z1_dim * kl_z1 + z2_dim * kl_z2)
```
Comment (martinjankowiak) [inline test/infer/test_gradient.py:610]: can we add a test where the scale only hits some sites and not all?

**Review by martinjankowiak — APPROVED** [review]

**Comment by martinjankowiak** [comment]
looks great @ordabayevy ! thanks for adding all these detailed tests. would be nice to get that additional `scale` test in here but i think this is basically ready to merge

**Comment by ordabayevy** [comment]
@fehiepsi  @martinjankowiak thanks for reviewing. @martinjankowiak I will just a second scale to that test. Should be easy to do.

**Comment by martinjankowiak** [comment]
@ordabayevy btw i imagine these added tests are pretty fast right?

**Comment by ordabayevy** [comment]
@martinjankowiak on my MacBook Pro they finish in 16s
